### PR TITLE
[TensorRT EP] Refactor of TRT plugins support

### PR DESCRIPTION
**Pull request by chilo-ms**
Make sure "trt.plugins" custom op domain only being registered once.
The bottom line is "trt.plugins" custom op domain needs to be registered before model load.

`CreateTensorRTCustomOpDomainList()` is TRT EP's function to create "trt.plugins" custom op domain. Following are places where this function will be called. (This function only fetches all the TRT plugins from TRT plugin registry but not yet registered them to ORT custom op registry. The real registration happens in AddCustomOpDomains())

C/C++ APIs:

- `OrtApis::SessionOptionsAppendExecutionProvider_TensorRT_XX`: This function will make session option object contain the "trt.plugins" custom op domain for ORT to register. So that later the session creation api can register the custom op domain accordingly and won't complain about invalid onnx node.
- `InferenceSession::RegisterExecutionProvider`: In some cases, users might create the session object first and later call session_object.RegisterExecutionProvider(). This function will call p_exec_provider->GetCustomOpDomainList() which returns "trt.plugins" custom op domain. Otherwise, session_object.Load(model) will complain. 

Python APIs:

- `RegisterTensorRTPluginsAsCustomOps`: Need to call this function so that session option object contains the "trt.plugins" custom op domain for ORT to register. 

Different language bindings have slightly different workflow of initializing the session. This might cause duplicate custom op domain in `session_option.custom_op_domains_` or `CreateTensorRTCustomOpDomainList()` being called more than once, but we put checks to make sure ep's custom op domain won't be registered twice.

### CODE REVIEW

**Comment by jywu-msft** [comment]
are there any unit tests we can add to enhance the coverage?

**Comment by chilo-ms** [comment]
> are there any unit tests we can add to enhance the coverage?

Will add a unit test for this

**Comment by chilo-ms** [comment]
unit tests added

File: onnxruntime/test/python/onnxruntime_test_python.py [hunk 1]
```diff
@@ -298,6 +298,14 @@ def test_set_providers_with_options(self):
             self.assertEqual(option["trt_engine_cache_path"], str(engine_cache_path))
             self.assertEqual(option["trt_force_sequential_engine_build"], "1")
 
+            from onnxruntime.capi import _pybind_state as C
+
+            session_options = C.get_default_session_options()
+
+            # TRT plugins registered as custom op domain should only be added once in session option regaldless of number of session creation
```
Comment (jywu-msft) [inline onnxruntime/test/python/onnxruntime_test_python.py:305]: typo: regardless.

suggest you actually keep the sessions around
 sess1 = onnxrt.InferenceSession(...)
and sess 2 = onnxrt.InferenceSession(...)

Comment (jywu-msft) [reply onnxruntime/test/python/onnxruntime_test_python.py:305]: just a thought: i guess there's no way to read back what ops/trtplugins were registered implicitly. maybe that would be a nice api to have?

Comment (chilo-ms) [reply onnxruntime/test/python/onnxruntime_test_python.py]: > typo: regardless.
> 
> suggest you actually keep the sessions around sess1 = onnxrt.InferenceSession(...) and sess 2 = onnxrt.InferenceSession(...)

Fixed

Comment (chilo-ms) [reply onnxruntime/test/python/onnxruntime_test_python.py]: > just a thought: i guess there's no way to read back what ops/trtplugins were registered implicitly. maybe that would be a nice api to have?

Currently there is no such API, but yes it's good to have. Probably can add it in another PR.

**Review by jywu-msft — APPROVED** [review]

**Review by jywu-msft — DISMISSED** [review]

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -613,9 +613,35 @@ common::Status InferenceSession::RegisterExecutionProvider(const std::shared_ptr
...
+  } else {
+    for (auto candidate_custom_op_domain : candidate_custom_op_domains) {
+      for (auto registry_kernel : registry_kernels) {
+        const auto& kernel_map = registry_kernel->GetKernelCreateMap();
+        bool need_resigter = true;
+        // If the kernel registry is the ep's custom op registry, we only need to check the first kernel,
+        // because all kernels in one kernel registry should have the same domain name.
+        for (auto iter = kernel_map.begin(); iter != kernel_map.end(); iter++) {
+          if (iter->second.kernel_def->Domain() == candidate_custom_op_domain->domain_) {
+            need_resigter = false;
```
Comment (jywu-msft) [inline onnxruntime/core/session/inference_session.cc:635]: typo register

File: onnxruntime/core/session/inference_session.cc [hunk 3]
```diff
@@ -613,9 +613,35 @@ common::Status InferenceSession::RegisterExecutionProvider(const std::shared_ptr
...
+  auto registry_kernels = kernel_registry_manager_.GetKernelRegistriesByProviderType(p_exec_provider->Type());
+
+  // Register the custom op domain only if it has not been registered before
+  if (registry_kernels.empty()) {
+    custom_op_domains = candidate_custom_op_domains;
+  } else {
+    for (auto candidate_custom_op_domain : candidate_custom_op_domains) {
+      for (auto registry_kernel : registry_kernels) {
+        const auto& kernel_map = registry_kernel->GetKernelCreateMap();
+        bool need_resigter = true;
```
Comment (jywu-msft) [inline onnxruntime/core/session/inference_session.cc:630]: typo register

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**